components/feature/autofill/src/main/java/mozilla/components/feature/autofill/verify/CredentialAccessVerifier.kt

```diff
@@ -0,0 +1,50 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.feature.autofill.verify
+
+import android.content.Context
+import mozilla.components.service.digitalassetlinks.AndroidAssetFinder
+import mozilla.components.service.digitalassetlinks.AssetDescriptor
+import mozilla.components.service.digitalassetlinks.Relation
+import mozilla.components.service.digitalassetlinks.local.StatementRelationChecker
+
+/**
+ * Helper to verify that a specific application is allowed to receive get the login credentials for
+ * a specific domain.
+ *
+ * The verification is done through Digital Asset Links, which allow a domain to specify associated
+ * apps and their signatures.
+ * - https://developers.google.com/digital-asset-links/v1/getting-started
+ * - https://github.com/google/digitalassetlinks/blob/master/well-known/details.md
+ */
+class CredentialAccessVerifier(
+    private val checker: StatementRelationChecker,
+    private val assetsFinder: AndroidAssetFinder = AndroidAssetFinder()
+) {
+    /**
+     * Verifies and returns `true` if the application with [packageName] is allowed to receive
+     * credentials for [domain] according to the hosted Digital Assets Links file. Returns `false`
+     * otherwise. This method may also return `false` if a verification could not be performed,
+     * e.g. the device is offline.
+     */
+    fun hasCredentialRelationship(
+        context: Context,
+        domain: String,
+        packageName: String
+    ): Boolean {
+        val assets = assetsFinder.getAndroidAppAsset(packageName, context.packageManager).toList()
+
+        // I was expecting us to need to verify all signatures here. But If I understand the usage
+        // in `OriginVerifier` and the spec (see link in class comment) correctly then verifying one
+        // certificate is enough to identify an app.
+        val asset = assets.firstOrNull() ?: return false
```

Comment:
I _think_ that's right, yeah. Comparing to the package signature handling code in AccountSharing, at least. https://github.com/mozilla-mobile/android-components/blob/1007c236003f5ac94bb3aa93e62b151e9ce14496/components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/sharing/AccountSharing.kt#L151 - that one's more paranoid (e.g. here it's probably fine to allow for multiple signers and signature rotation, i don't think that has a bearing on the check here).

My only vague concern is how this will handle a case of multiple signers, but it's not clear to me Play Store actually supports that...

Reply:
There can be multiple signers. In this case we'd pick one signature and the code that fetches from package manager creates multiple "Android App Assets" (one per signature) and we look for one that matches. So we'd accept if at least one signature matches.

My concern here was about signing key leaks. But I guess in that situation the leaked key would get removed from the digital asset links file too.